# Green Chemistry Lecture: 3. Process Mass Intensity (PMI) Calculator

## Principle 1: Waste Prevention & Systems Thinking

**Process Mass Intensity (PMI)** is a comprehensive metric championed by the ACS Green Chemistry Institute Pharmaceutical Roundtable. It evaluates the efficiency of the *entire process* by comparing the total mass of all materials used to the mass of the final product.

**The Goal:** Minimize PMI. An ideal PMI is 1, meaning only the reactants' mass contributes to the product's mass, with no solvents or other waste. This is practically impossible, but it serves as a useful benchmark.

### Formula
$$\text{PMI} = \frac{\text{Total Mass of All Inputs (kg)}}{\text{Mass of Product (kg)}}$$

### How is PMI different from E-Factor?

They are closely related! They both measure waste, but from different perspectives.

-   **PMI** focuses on the **inputs**. It asks: "How much stuff did I use to make my product?"
-   **E-Factor** focuses on the **outputs**. It asks: "How much waste did I make?"

The relationship is simple: **PMI = E-Factor + 1**

PMI is often preferred in an industrial setting because it's easier to track the mass of materials purchased and brought into a factory than it is to precisely measure all waste streams leaving it.

### Example: A Pharmaceutical Intermediate Synthesis

This example models a more industrial-style process, where detailed tracking of all materials is crucial. We define the mass for every single component used in the reaction, workup, and final purification (crystallization).

* **Inputs:** A Python dictionary, `inputs`, is used to store the name and mass (in kg) of every material used in the process.
* **Output:** The `final_product_mass` variable holds the mass of the isolated product.

### Code Explanation

The `calculate_pmi` function is designed to provide a detailed breakdown of material usage:

1.  **Total Mass Calculation:** It first calculates the `total_input_mass` by summing all values in the `inputs` dictionary.
2.  **PMI and E-Factor:** It computes both the PMI and the related E-Factor.
3.  **Detailed Breakdown:** A key feature of the function is that it loops through the `inputs` dictionary to print each material's mass alongside its **percentage contribution** to the total mass. This immediately highlights which materials are the most significant drivers of the process's overall mass intensity.

In [1]:
def calculate_pmi(product_mass: float, input_materials: dict):
    """
    Calculates the Process Mass Intensity (PMI) for a chemical process.
    
    Args:
        product_mass (float): The actual isolated mass of the desired product (in kg).
        input_materials (dict): A dictionary where keys are material names and values are their masses (in kg).
    
    Returns:
        None. Prints the results directly.
    """
    if product_mass <= 0:
        print("Error: Product mass must be greater than zero.")
        return

    total_input_mass = sum(input_materials.values())
    pmi = total_input_mass / product_mass
    e_factor = pmi - 1

    # --- Display Results ---
    print("--- Process Mass Intensity (PMI) Calculation ---")
    print(f"Mass of Isolated Product: {product_mass:.3f} kg")
    print("\n--- Input Materials Breakdown ---")
    for material, mass in input_materials.items():
        percentage = (mass / total_input_mass) * 100
        print(f"  - {material:<20}: {mass:>7.3f} kg ({percentage:.1f}% of total input)")
    
    print("-" * 40)
    print(f"Total Mass of All Inputs: {total_input_mass:.3f} kg")
    print("\n" + "="*40)
    print(f"Process Mass Intensity (PMI): {pmi:.2f}")
    print(f"(Related E-Factor: {e_factor:.2f})")
    print("="*40)
    print(f"\nInterpretation: {pmi:.2f} kg of material were used for every 1 kg of product made.")

# --- Example: A multi-step process for a pharmaceutical intermediate ---
# All masses are in kg.

inputs = {
    "Starting Material A": 10.0,
    "Reagent B": 12.5,
    "Catalyst": 0.1,
    "Toluene (Solvent)": 150.0,
    "HCl (Workup)": 25.0,
    "Water (Workup)": 200.0,
    "Sodium Bicarbonate": 30.0,
    "Heptane (Crystallization)": 80.0
}

# After the full process, we isolate 18.5 kg of the final product.
final_product_mass = 18.5

# Note: For this calculation, we are including water.
calculate_pmi(product_mass=final_product_mass, input_materials=inputs)

--- Process Mass Intensity (PMI) Calculation ---
Mass of Isolated Product: 18.500 kg

--- Input Materials Breakdown ---
  - Starting Material A :  10.000 kg (2.0% of total input)
  - Reagent B           :  12.500 kg (2.5% of total input)
  - Catalyst            :   0.100 kg (0.0% of total input)
  - Toluene (Solvent)   : 150.000 kg (29.6% of total input)
  - HCl (Workup)        :  25.000 kg (4.9% of total input)
  - Water (Workup)      : 200.000 kg (39.4% of total input)
  - Sodium Bicarbonate  :  30.000 kg (5.9% of total input)
  - Heptane (Crystallization):  80.000 kg (15.8% of total input)
----------------------------------------
Total Mass of All Inputs: 507.600 kg

Process Mass Intensity (PMI): 27.44
(Related E-Factor: 26.44)

Interpretation: 27.44 kg of material were used for every 1 kg of product made.


### Conclusion and Interpretation
The calculated **Process Mass Intensity (PMI) of 27.44** provides a holistic view of the process's material efficiency and clearly demonstrates the value of systems thinking in green chemistry.

1. **Identifying Mass Inefficiency**: A PMI of 27.44 reveals that for every 1 kg of final product, over 27 kg of materials were consumed. The value-added starting materials and reagents (Starting Material A and Reagent B) constitute only 4.5% of the total mass brought into the process.
2. **Pinpointing the Source of Waste**: The input materials breakdown is the most powerful feature of the PMI analysis. It shows unequivocally that auxiliary materials are the primary drivers of inefficiency:
- **Water (Workup)** is the single largest component at 39.4%.
- **Solvents (Toluene and Heptane)** combined account for 45.4% of the total mass.
- **Together**, water and organic solvents make up nearly 85% of the total process mass.

**Final Takeaway**: This analysis perfectly illustrates why PMI is a critical metric for industrial green chemistry. While chemists often focus on optimizing reaction yield, the PMI calculation proves that the most significant environmental (and economic) gains are often achieved by focusing on the entire system. By targeting the massive solvent and water volumes used during reaction and purification, chemists and engineers can make a much larger impact on waste prevention than by simply increasing the reaction yield by a few percentage points.